### DISCUSSION TITLE
Force reverse cursor question

### DISCUSSION THREAD

**Discussion by gwww** [wezterm/wezterm#2861] [General]
I'm using `config.force_reverse_video_cursor = true`. The reverse cursor seems to be no so reverse in some cases. In the video below when the cursor in on the LHS of the screen it seems to take on the colour of the indent line. Similarly when under the "i" in the if. I don't know if this is the expected behaviour. The app running is neovim.

https://user-images.githubusercontent.com/3060128/207492958-ac1ce22a-da7a-4c9c-bd8d-191ded97e23a.mov

**Comment by gwww** [comment]
I tried using the same neovim and file with iTerm2. It seems better, at least on the indent lines. I can post video if that helps.

**Comment by wez** [comment]
The indent line is just a unicode character consisting of a vertical line (likely https://www.fileformat.info/info/unicode/char/258f/index.htm) that your editor has printed there, so it seems totally reasonable for the cursor to reverse out the way it does over that glyph.

**Reply by gwww** [reply]
Does it look that way because the cursor is actually a box and a border and each of those two things are independently inverting?

**Reply by wez** [reply]
The cursor is rendered with the foreground color of the cell when in reverse video mode, and the text is rendered in the background color, similar to having the reverse video attribute set.

**Comment by gwww** [comment]
Thank @wez! I now understand, its a simple reverse fg and bg colours. The problem being when fg and bg colors are very similar. For a first dumb cut at a smart cursor, would doing something such the following be better?

```
if abs(fg-bg) < some_smallish_value then    # where small == around 20% of #ffffff
    cursor_bg = #ffffff - min(fg,bg)
endif
```

**Comment by gwww** [comment]
I started to wonder what other terminal emulators do for a low contrast cursor situation. I see (by using) iTerm2 that it does something but I could not find the code or docs to explain.

I decided to look further. From https://sw.kovidgoyal.net/kitty/conf/#opt-kitty.cursor, Kitty does the following:
> Also, the cursor colors are modified if the cell background and foreground colors have very low contrast.

Code here: https://github.com/kovidgoyal/kitty/blob/fda4aa21a1d29acf3c74ea82dace740120b1a387/kitty/shaders.c#L274-L287. The referenced function `rgb_contrast` here: https://github.com/kovidgoyal/kitty/blob/c48049a3bb727d2a9135cfb2ef0d89a8c349f1f7/kitty/colors.h#L65-L76

**Reply by wez** [reply]
I'm not opposed to having wezterm do some color math to measure contrast and adjust, if you wanted to put together a PR.

https://github.com/wez/wezterm/blob/main/wezterm-gui/src/termwindow/render.rs#L3447-L3478 is where we currently check for `force_reverse_video_cursor`.

`SrgbaTuple` has a method for computing the contrast ratio:
https://github.com/wez/wezterm/blob/main/color-types/src/lib.rs#L482

but the render logic operates on linearized color:
https://github.com/wez/wezterm/blob/main/color-types/src/lib.rs#L762

you could make a quick first pass by doing something like `params.bg_color.to_srgb().contrast_ratio(params.fg_color.to_srgb())`, but in a final version we'd want to avoid those srgb conversions because the cell attributes are computed for every cell and that aggregates to a lot of excess math per frame.  So in short: we'd want a version of luminance calculation that works with linear rgb, and it maybe as simple as adding a method to `LinearRgba` based on the functions you found in kitty.
